internal/resources/cluster/data_source_cluster.go

```diff
@@ -129,18 +129,28 @@ func dataSourceTMCClusterRead(ctx context.Context, d *schema.ResourceData, m int
 
 	clusterSpec := constructSpec(d)
 
-	switch {
-	case resp.Cluster.Spec.TkgAws != nil:
-		resp.Cluster.Spec.TkgAws.Topology.NodePools = clusterSpec.TkgAws.Topology.NodePools
-	case resp.Cluster.Spec.TkgServiceVsphere != nil:
-		resp.Cluster.Spec.TkgServiceVsphere.Topology.NodePools = clusterSpec.TkgServiceVsphere.Topology.NodePools
-	case resp.Cluster.Spec.TkgVsphere != nil:
-		resp.Cluster.Spec.TkgVsphere.Topology.NodePools = clusterSpec.TkgVsphere.Topology.NodePools
-	}
+	getNodepoolForCluster(resp.Cluster.Spec, clusterSpec)
 
 	if err := d.Set(SpecKey, flattenSpec(resp.Cluster.Spec)); err != nil {
 		return diag.FromErr(err)
 	}
 
 	return diags
 }
+
+func getNodepoolForCluster(respSpec, clusterSpec *clustermodel.VmwareTanzuManageV1alpha1ClusterSpec) {
+	if clusterSpec == nil {
+		return
+	}
+
+	switch {
+	case clusterSpec.TkgAws == nil && clusterSpec.TkgServiceVsphere == nil && clusterSpec.TkgVsphere == nil:
+		break
+	case respSpec.TkgAws != nil && respSpec.TkgAws.Topology != nil && clusterSpec.TkgAws != nil && clusterSpec.TkgAws.Topology != nil:
```

Comment:
If respSpec.TkgAWS is nil and if TkgAWS needs to populated, then we should initialise the struct and assign the value.